File: Assemblies/Protean.Tools/FileHelper.vb
Imports System
Imports System.Text
Imports System.IO

'' ---------------------------------------------------------
'' Class       :   Eonic.Tools.FileHelper
'' Author      :   Ali Granger
'' Website     :   www.eonic.co.uk
'' Description :   File classes and functions.
''
'' (c) Copyright 2010 Eonic Ltd.
'' ---------------------------------------------------------

Public Class FileHelper




    Shared Function ReplaceIllegalChars(ByVal filePath As String, Optional ByVal replacementChar As String = "-") As String

        Try

            ' Replace illegal characters 
            For Each illegalChar As Char In Path.GetInvalidFileNameChars()
                filePath = filePath.Replace(illegalChar.ToString(), replacementChar)
            Next

            For Each illegalChar As Char In Path.GetInvalidPathChars()
                filePath = filePath.Replace(illegalChar.ToString(), replacementChar)
            Next

            Return filePath

        Catch ex As Exception
            Return ""
        End Try

    End Function


    ''' <summary>
    ''' Returns a MIME type for a given extension
    ''' </summary>
    ''' <param name="extension"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Shared Function GetMIMEType(ByVal extension As String) As String

        Dim mimeType As String

        Select Case extension.ToUpper
            'Common documents
            Case "TXT", "TEXT", "JS", "VBS", "ASP", "CGI", "PL", "NFO", "ME", "DTD"
                mimeType = "text/plain"
            Case "HTM", "HTML", "HTA", "HTX", "MHT"
                mimeType = "text/html"
            Case "CSV"
                mimeType = "text/comma-separated-values"
            Case "JS"
                mimeType = "text/javascript"
            Case "CSS"
                mimeType = "text/css"
            Case "PDF"
                mimeType = "application/pdf"
            Case "RTF"
                mimeType = "application/rtf"
            Case "XML", "XSL", "XSLT"
                mimeType = "text/xml"
            Case "WPD"
                mimeType = "application/wordperfect"
            Case "WRI"
                mimeType = "application/mswrite"
            Case "XLS", "XLS3", "XLS4", "XLS5", "XLW"
                mimeType = "application/msexcel"
            Case "DOC"
                mimeType = "application/msword"
            Case "PPT", "PPS"
                mimeType = "application/mspowerpoint"

                'WAP/WML files 
            Case "WML"
                mimeType = "text/vnd.wap.wml"
            Case "WMLS"
                mimeType = "text/vnd.wap.wmlscript"
            Case "WBMP"
                mimeType = "image/vnd.wap.wbmp"
            Case "WMLC"
                mimeType = "application/vnd.wap.wmlc"
            Case "WMLSC"
                mimeType = "application/vnd.wap.wmlscriptc"

                'Images
            Case "GIF"
                mimeType = "image/gif"
            Case "JPG", "JPE", "JPEG"
                mimeType = "image/jpeg"
            Case "PNG"
                mimeType = "image/png"
            Case "BMP"
                mimeType = "image/bmp"
            Case "TIF", "TIFF"
                mimeType = "image/tiff"
            Case "AI", "EPS", "PS"
                mimeType = "application/postscript"

                'Sound files
            Case "AU", "SND"
                mimeType = "audio/basic"
            Case "WAV"
                mimeType = "audio/wav"
            Case "RA", "RM", "RAM"
                mimeType = "audio/x-pn-realaudio"
            Case "MID", "MIDI"
                mimeType = "audio/x-midi"
            Case "MP3"
                mimeType = "audio/mp3"
            Case "M3U"
                mimeType = "audio/m3u"

                'Video/Multimedia files
            Case "ASF"
                mimeType = "video/x-ms-asf"
            Case "AVI"
                mimeType = "video/avi"
            Case "MPG", "MPEG"
                mimeType = "video/mpeg"
            Case "QT", "MOV", "QTVR"
                mimeType = "video/quicktime"
            Case "SWA"
                mimeType = "application/x-director"
            Case "SWF"
                mimeType = "application/x-shockwave-flash"
                'Compressed/archives
            Case "ZIP"
                mimeType = "application/x-zip-compressed"
            Case "GZ"
                mimeType = "application/x-gzip"
            Case "RAR"
                mimeType = "application/x-rar-compressed"

                'Miscellaneous
            Case "COM", "EXE", "DLL", "OCX"
                mimeType = "application/octet-stream"

                'Unknown (send as binary stream)
            Case Else
                mimeType = "application/octet-stream"
        End Select

        Return mimeType

    End Function



End Class

File: Assemblies/Protean.Tools_Net7/FileHelper.cs
using System;
using System.IO;

namespace Protean.Tools
{

    public class FileHelper
    {

        public static string ReplaceIllegalChars(string filePath, string replacementChar = "-")
        {

            try
            {

                // Replace illegal characters 
                foreach (char illegalChar in Path.GetInvalidFileNameChars())
                    filePath = filePath.Replace(illegalChar.ToString(), replacementChar);

                foreach (char illegalChar in Path.GetInvalidPathChars())
                    filePath = filePath.Replace(illegalChar.ToString(), replacementChar);

                return filePath;
            }

            catch (Exception ex)
            {
                return "";
            }

        }


        /// <summary>
        /// Returns a MIME type for a given extension
        /// </summary>
        /// <param name="extension"></param>
        /// <returns></returns>
        /// <remarks></remarks>
        public static string GetMIMEType(string extension)
        {

            string mimeType;

            switch (extension.ToUpper() ?? "")
            {
                // Common documents
                case "TXT":
                case "TEXT":
                case "JS":
                case "VBS":
                case "ASP":
                case "CGI":
                case "PL":
                case "NFO":
                case "ME":
                case "DTD":
                    {
                        mimeType = "text/plain";
                        break;
                    }
                case "HTM":
                case "HTML":
                case "HTA":
                case "HTX":
                case "MHT":
                    {
                        mimeType = "text/html";
                        break;
                    }
                case "CSV":
                    {
                        mimeType = "text/comma-separated-values";
                        break;
                    }
                case var @case when @case == "JS":
                    {
                        mimeType = "text/javascript";
                        break;
                    }
                case "CSS":
                    {
                        mimeType = "text/css";
                        break;
                    }
                case "PDF":
                    {
                        mimeType = "application/pdf";
                        break;
                    }
                case "RTF":
                    {
                        mimeType = "application/rtf";
                        break;
                    }
                case "XML":
                case "XSL":
                case "XSLT":
                    {
                        mimeType = "text/xml";
                        break;
                    }
                case "WPD":
                    {
                        mimeType = "application/wordperfect";
                        break;
                    }
                case "WRI":
                    {
                        mimeType = "application/mswrite";
                        break;
                    }
                case "XLS":
                case "XLS3":
                case "XLS4":
                case "XLS5":
                case "XLW":
                    {
                        mimeType = "application/msexcel";
                        break;
                    }
                case "DOC":
                    {
                        mimeType = "application/msword";
                        break;
                    }
                case "PPT":
                case "PPS":
                    {
                        mimeType = "application/mspowerpoint";
                        break;
                    }

                // WAP/WML files 
                case "WML":
                    {
                        mimeType = "text/vnd.wap.wml";
                        break;
                    }
                case "WMLS":
                    {
                        mimeType = "text/vnd.wap.wmlscript";
                        break;
                    }
                case "WBMP":
                    {
                        mimeType = "image/vnd.wap.wbmp";
                        break;
                    }
                case "WMLC":
                    {
                        mimeType = "application/vnd.wap.wmlc";
                        break;
                    }
                case "WMLSC":
                    {
                        mimeType = "application/vnd.wap.wmlscriptc";
                        break;
                    }

                // Images
                case "GIF":
                    {
                        mimeType = "image/gif";
                        break;
                    }
                case "JPG":
                case "JPE":
                case "JPEG":
                    {
                        mimeType = "image/jpeg";
                        break;
                    }
                case "PNG":
                    {
                        mimeType = "image/png";
                        break;
                    }
                case "BMP":
                    {
                        mimeType = "image/bmp";
                        break;
                    }
                case "TIF":
                case "TIFF":
                    {
                        mimeType = "image/tiff";
                        break;
                    }
                case "AI":
                case "EPS":
                case "PS":
                    {
                        mimeType = "application/postscript";
                        break;
                    }

                // Sound files
                case "AU":
                case "SND":
                    {
                        mimeType = "audio/basic";
                        break;
                    }
                case "WAV":
                    {
                        mimeType = "audio/wav";
                        break;
                    }
                case "RA":
                case "RM":
                case "RAM":
                    {
                        mimeType = "audio/x-pn-realaudio";
                        break;
                    }
                case "MID":
                case "MIDI":
                    {
                        mimeType = "audio/x-midi";
                        break;
                    }
                case "MP3":
                    {
                        mimeType = "audio/mp3";
                        break;
                    }
                case "M3U":
                    {
                        mimeType = "audio/m3u";
                        break;
                    }

                // Video/Multimedia files
                case "ASF":
                    {
                        mimeType = "video/x-ms-asf";
                        break;
                    }
                case "AVI":
                    {
                        mimeType = "video/avi";
                        break;
                    }
                case "MPG":
                case "MPEG":
                    {
                        mimeType = "video/mpeg";
                        break;
                    }
                case "QT":
                case "MOV":
                case "QTVR":
                    {
                        mimeType = "video/quicktime";
                        break;
                    }
                case "SWA":
                    {
                        mimeType = "application/x-director";
                        break;
                    }
                case "SWF":
                    {
                        mimeType = "application/x-shockwave-flash";
                        break;
                    }
                // Compressed/archives
                case "ZIP":
                    {
                        mimeType = "application/x-zip-compressed";
                        break;
                    }
                case "GZ":
                    {
                        mimeType = "application/x-gzip";
                        break;
                    }
                case "RAR":
                    {
                        mimeType = "application/x-rar-compressed";
                        break;
                    }

                // Miscellaneous
                case "COM":
                case "EXE":
                case "DLL":
                case "OCX":
                    {

                        // Unknown (send as binary stream)
                        mimeType = "application/octet-stream";
                        break;
                    }

                default:
                    {
                        mimeType = "application/octet-stream";
                        break;
                    }
            }

            return mimeType;

        }



    }
}
